docs/enhancements/20190724-prometheus-refactoring.md

```diff
@@ -0,0 +1,370 @@
+
+## Motivation
+
+The current implementation lacks enough configuration options to allow the customization of the features to use in the metrics. Also, due to the lack of iterations to improve metrics, the current state has several issues:
+
+- Lack of details about endpoints
+- Performance issues in scenarios with high RPS
+- Use of histogram and summaries introduce high cardinality
+- High cardinality of metrics in general.
+
+### Goals
+
+- Customizable prometheus metrics for any workload without compromising NGINX performance.
+- Metrics counter are initialized at start/reload [#4066](https://github.com/kubernetes/ingress-nginx/issues/4066)
+- Removal of flags, moving these flags to the configuration configmap. The reason for this is to not require a change in the ingress controller deployment.
+- Expose number of endpoints per ingress via prometheus [#3899](https://github.com/kubernetes/ingress-nginx/issues/3899)
+- Fix several issues:
+
+  - https://github.com/kubernetes/ingress-nginx/issues/3936
+  - https://github.com/kubernetes/ingress-nginx/issues/4026
+  - https://github.com/kubernetes/ingress-nginx/issues/3645
+  - https://github.com/kubernetes/ingress-nginx/issues/3818
+  - https://github.com/kubernetes/ingress-nginx/issues/3713
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
+  - https://github.com/kubernetes/ingress-nginx/issues/2924
+  - https://github.com/kubernetes/ingress-nginx/issues/3898
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
+
+### Non-Goals
+
+Update of the Grafana dashboard [provided by the project](https://github.com/kubernetes/ingress-nginx/tree/master/deploy/grafana/dashboards). This should be done after the implementation of this KEP.
+
+## Proposal
+
+Test if using message pack we can improve the performance of the Lua communication with the ingress controller ([lua library](https://luarocks.org/modules/antirez/lua-cmsgpack))
+
+One of the issues we have is the range of options and configuration users expected from the exposed metrics. This creates a maintenance problem due to the fact the metrics are hardcoded in the source code.
+For this reason, one of the goals in this KEP is to introduce a CRD for metrics, allowing users to define which rules they want to use and which labels should be used.
+As an example:
+
+```yaml
+kind: metric
+metadata:
+  name: request_duration_seconds
+  namespace: default
+spec:
+  help: The request processing time in milliseconds
+  constLabels:
+    - controller_pod
+  labels:
+    - ingress_uid # label to join with the ingress info metric using group_left
+    - path        # same ^^
+    - method
+    - status
```

Comment:
An additional issue #3818 can be solved by incorporating a status_class:5xx in the metric implemented as part of this proposal